[Site Isolation] Fix `WKAccessibilityRootObject`

charliewolfe opened the pull request:
#### 833adfcdc125889ba145a1d95077dcf398d59f8b
<pre>
[Site Isolation] Fix `WKAccessibilityRootObject`
<a href="https://rdar.apple.com/132428902">rdar://132428902</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=277028">https://bugs.webkit.org/show_bug.cgi?id=277028</a>

Reviewed by Alex Christensen.

`WKAccessibilityRootObject` tries to get the root AX object on the main frame&apos;s document, which won&apos;t
work if the main frame is out-of-process. Add `_WKAccessibilityRootObject` to get the root AX object
from the frame that is calling `accessibleElementById`.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKAccessibilityRootObject):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFramePrivate.h:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKAccessibilityRootObject): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.cpp:
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::elementAtPoint):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.h:
* Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityControllerAtspi.cpp:
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::focusedElement):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityControllerIOS.mm:
(WTR::AccessibilityController::focusedElement):
(WTR::AccessibilityController::accessibleElementById):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityControllerMac.mm:
(WTR::AccessibilityController::focusedElement):
(WTR::AccessibilityController::accessibleElementById):
* Tools/WebKitTestRunner/InjectedBundle/playstation/AccessibilityControllerPlayStation.cpp:
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::focusedElement):
* Tools/WebKitTestRunner/InjectedBundle/win/AccessibilityControllerWin.cpp:
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::focusedElement):

Canonical link: <a href="https://commits.webkit.org/281333@main">https://commits.webkit.org/281333@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4efbdf30cc937653c1c4dd3c6e931ee1e1c24c24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59534 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38879 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12058 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63449 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10052 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61663 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46532 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10209 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48316 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7046 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61564 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36298 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51535 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29149 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33003 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8788 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8981 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54951 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9066 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65181 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3462 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8960 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55657 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3473 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51528 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55774 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2876 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8894 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34693 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35776 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36862 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35521 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
```diff
@@ -299,3 +299,29 @@ void _WKBundleFrameGenerateTestReport(WKBundleFrameRef frameRef, WKStringRef mes
+    if (!frame)
+        return nullptr;
+
+    while (auto* parent = dynamicDowncast<WebCore::LocalFrame>(frame->tree().parent()))
```
LocalFrame::rootFrame can be called instead of this.  We store a pointer in order to avoid loops like this.

achristensen07 commented on Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityControllerMac.mm:
```diff
@@ -132,10 +133,9 @@ static id findAccessibleObjectById(id obj, NSString *idAttribute)
 {
-    auto page = InjectedBundle::singleton().page()->page();
-    PlatformUIElement root = static_cast<PlatformUIElement>(WKAccessibilityRootObject(page));
+    id root = static_cast<PlatformUIElement>(_WKAccessibilityRootObject(WKBundleFrameForJavaScriptContext(context)));
```
Why don't we make the variable type a PlatformUIElement instead of an id?

achristensen07 commented on Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityControllerIOS.mm:
```diff
@@ -99,10 +100,9 @@ static id findAccessibleObjectById(id obj, NSString *idAttribute)
+RefPtr<AccessibilityUIElement> AccessibilityController::accessibleElementById(JSContextRef context, JSStringRef idAttribute)
 {
-    auto page = InjectedBundle::singleton().page()->page();
-    id root = static_cast<PlatformUIElement>(WKAccessibilityRootObject(page));
```
We should remove or at least deprecate WKAccessibilityRootObject.  I don't think there are any non-WebKitTestRunner users, so we should be able to remove it.

achristensen07 commented on Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFramePrivate.h:
```diff
@@ -57,6 +57,8 @@ WK_EXPORT void WKBundleFrameFocus(WKBundleFrameRef frame);
 
 WK_EXPORT void _WKBundleFrameGenerateTestReport(WKBundleFrameRef, WKStringRef message, WKStringRef group);
 
+WK_EXPORT void* _WKAccessibilityRootObject(WKBundleFrameRef frame);
```
It would be nice if we could use the right type instead of void* somehow, but void* is the status quo, so probably not important for this PR.

Also, the underscore is probably not necessary.  It's a convention for ObjC that isn't public API, but this is C and all the bundle isn't public API.

achristensen07 commented on LayoutTests/platform/mac-site-isolation/TestExpectations:
```diff
@@ -372,6 +372,7 @@ accessibility/activation-of-input-field-inside-other-element.html [ Skip ]
 accessibility/add-children-pseudo-element.html [ Skip ]
 accessibility/ancestor-computation.html [ Skip ]
 accessibility/anonymous-render-block-in-continuation-causes-crash.html [ Skip ]
+accessibility/aria-controls-with-tabs.html [ Timeout ]
```
Can we skip instead of mark as an expected timeout?  That'll help tests go faster, and that's what we do for other tests.

charliewolfe committed:
[Site Isolation] Fix `WKAccessibilityRootObject`
rdar://132428902
https://bugs.webkit.org/show_bug.cgi?id=277028

Reviewed by Alex Christensen.

`WKAccessibilityRootObject` tries to get the root AX object on the main frame's document, which won't
work if the main frame is out-of-process. Add `_WKAccessibilityRootObject` to get the root AX object
from the frame that is calling `accessibleElementById`.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKAccessibilityRootObject):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFramePrivate.h:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKAccessibilityRootObject): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.cpp:
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::elementAtPoint):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.h:
* Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityControllerAtspi.cpp:
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::focusedElement):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityControllerIOS.mm:
(WTR::AccessibilityController::focusedElement):
(WTR::AccessibilityController::accessibleElementById):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityControllerMac.mm:
(WTR::AccessibilityController::focusedElement):
(WTR::AccessibilityController::accessibleElementById):
* Tools/WebKitTestRunner/InjectedBundle/playstation/AccessibilityControllerPlayStation.cpp:
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::focusedElement):
* Tools/WebKitTestRunner/InjectedBundle/win/AccessibilityControllerWin.cpp:
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::focusedElement):

Canonical link: https://commits.webkit.org/281333@main
